### PR TITLE
Demo TF image classification

### PR DESCRIPTION
**Pull request by amyeroberts**
# What does this PR do?

Add a notebook that demonstrates how to do image classification with Hugging Face tensorflow models. 

TF implementation of: https://github.com/huggingface/notebooks/blob/main/examples/image_classification.ipynb

### CODE REVIEW

**Comment by amyeroberts** [comment]
cc @sayakpaul

File: examples/image_classification-tf.ipynb [hunk 1]
```diff
@@ -0,0 +1,7883 @@
+{
```
Comment (sayakpaul) [inline examples/image_classification-tf.ipynb:1]: <p>Should we also mention ViT MAE since it differs from the rest of the models its pre-training strategy (it's a self-supervised model)? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/notebooks/pull/202/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (sayakpaul) [inline examples/image_classification-tf.ipynb:1]: <p>Can we show a collage of the preprocessed samples? Or we'll just leave it to keep it inline with PyTorch example? It's fine if that's the case.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/notebooks/pull/202/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='42'/>

Comment (sayakpaul) [inline examples/image_classification-tf.ipynb:1]: <p>It's probably worth mentioning that the samples of <code>train_ds</code> and <code>val_ds</code>  have two keys - <code>pixel_values</code> and <code>label</code></p><p> </p><p>These <code>label</code>are used by HF to compute the model's loss internally.   </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/notebooks/pull/202/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='50'/>

Comment (sayakpaul) [inline examples/image_classification-tf.ipynb:1]: <p>Tensorflow -&gt; TensorFlow</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/notebooks/pull/202/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='52'/>

Comment (sayakpaul) [inline examples/image_classification-tf.ipynb:1]: <p>Would be interesting to benchmark the performance of this pipeline and a pure <code>tf.data</code> pipeline (I understand they might overlap here). </p><p> </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/notebooks/pull/202/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='54'/>

Comment (sayakpaul) [inline examples/image_classification-tf.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #4.        columns=["image", "label"],
</pre><p> </p><p>What is the significance of <code>columns</code>? Each sample of <code>train_ds</code> and <code>val_ds</code> has two keys: <code>pixel_values</code> and <code>label</code>.Will <code>to_tf_dataset</code> convert the <code>pixel_values</code> key to <code>image</code>? Or how is it?           </p><p> </p><p>Or better yet -- take a batch from <code>train_set</code> and investigate it (like printing the shape and so on).  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/notebooks/pull/202/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='54'/>

Comment (sayakpaul) [inline examples/image_classification-tf.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #7.    push_to_hub_model_id = f"{model_name}-finetuned-eurosat"
</pre><p> </p><p>Is it missing the <code>username</code> variable here? Or <code>notebook_login()</code> will automatically take care of this?   </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/notebooks/pull/202/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='58'/>

Comment (sayakpaul) [inline examples/image_classification-tf.ipynb:1]: <p>Why there's no <code>val_accuracy</code>?  <span class="ql-cursor">﻿</span></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/notebooks/pull/202/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='60'/>

Comment (sayakpaul) [inline examples/image_classification-tf.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #5.        epochs=3,
</pre><p> </p><p>Following the rest of the notebook, let's assign this value in a separate variable? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/notebooks/pull/202/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='60'/>

Comment (sayakpaul) [inline examples/image_classification-tf.ipynb:1]: <p>Will it make sense to also show how the users can use the model with their own explicit <code>tf.data</code> pipelines?  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/notebooks/pull/202/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='77'/>

**Comment by sayakpaul** [comment]
Great work @amyeroberts!

File: examples/image_classification-tf.ipynb (hunk 1 above)
Comment (amyeroberts) [reply examples/image_classification-tf.ipynb:1]: <p><code>notebook_login()</code> takes care of it, but you can also specify it e.g. <code>hub_model_id='username/model-checkpoint'</code> . You'll need to have write permission to push to the specified repo name/username</p>

Comment (amyeroberts) [reply examples/image_classification-tf.ipynb:1]: <p>I don't think ViT MAE has a classification head? As we're using <code>TFAutoModelForImageClassification</code>, we can only load models specified in the <code>TF_MODEL_FOR_IMAGE_CLASSIFICATION_MAPPING</code>   </p>

Comment (amyeroberts) [reply examples/image_classification-tf.ipynb:1]: <p>Yes - definitely! I think that work should be added in a separate PR (and possibly even notebook)</p>

Comment (amyeroberts) [reply examples/image_classification-tf.ipynb:1]: <p>Good idea! We can definitely add that. </p>

Comment (amyeroberts) [reply examples/image_classification-tf.ipynb:1]: <p><code>columns</code> specifies the columns to be read into the <code>tf.data.Dataset</code> object being created. (c.f. the <a href="https://huggingface.co/docs/datasets/v2.3.2/en/package_reference/main_classes#datasets.Dataset.to_tf_dataset" target="_blank">datasets</a> docs). <code>train_ds</code>  and <code>val_ds</code> don't have the <code>pixel_values</code>  key yet - they have <code>label</code> and <code>image </code> . <code>pixel_values</code> feature is added in the <code>preprocess_xxx</code> function.</p><p> </p><p> I'll add a cell below taking a batch and showing its contents 👍</p>

Comment (NielsRogge) [reply examples/image_classification-tf.ipynb:1]: ViTMAE directly fits into `ViTForImageClassification`.

Comment (sayakpaul) [reply examples/image_classification-tf.ipynb:1]: <p>Okay. </p><p> </p><p>Sorry for another follow-up question. If <code>image</code> and <code>label</code> are being loaded in the <code>tf.data.Dataset</code> object created how <code>pixel_values</code> is being used?     </p>

Comment (sayakpaul) [inline examples/image_classification-tf.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #78.        image = tf.image.crop_to_bounding_box(
</pre><p>How about following the standard transformation chain for validation images? </p><p> </p><p>Resizing (256 x 256) -&gt; Central Cropping (with an area factor of 0.875 which amounts to 224 x 224)?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/notebooks/pull/202/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='36'/>

Comment (sayakpaul) [reply examples/image_classification-tf.ipynb:1]: Cc @Rocketknight1

Comment (amyeroberts) [reply examples/image_classification-tf.ipynb:1]: <p>As I understand it, we can load VitMAE weights into a ViT model with <code>model = TFVitForImageClassification.from_pretrained(vit_mae_checkpoint)</code>  but there isn't a <code>TFVitMAEForImageClassification</code> object and <code>TFAutoModelForImageClassification.from_pretrained(vit_mae_checkpoint)</code>  throws an error because the model type isn't specified in <code>TF_MODEL_FOR_IMAGE_CLASSIFICATION_MAPPING</code>   (at least when I try). As a VitMAE checkpoint can't be used with <code>TFAutoModelForImageClassification</code> I didn't want to cause confusion and include it. </p><p> </p><p>I can mention and add a clarifying note?</p>

Comment (amyeroberts) [reply examples/image_classification-tf.ipynb:1]: <p>Comparing pipelines is definitely interesting and something to think about :) I'm not sure it belongs here as these notebooks are demoing a pipeline of how to do X and not necessarily comparing. </p>

Comment (sayakpaul) [reply examples/image_classification-tf.ipynb:1]: I'm sorry -- I should have been clearer. I just left that comment as a thought. The notebook absolutely doesn't need to include that.

Comment (Rocketknight1) [reply examples/image_classification-tf.ipynb:1]: <p>I suspect pure tf.data will be a lot faster right now, but we're working on that!</p>

Comment (Rocketknight1) [reply examples/image_classification-tf.ipynb:1]: <p>I would also have thought you'd need to put <code>pixel_values</code> in <code>columns</code>! I don't know why it isn't being dropped - let me investigate that.</p>

Comment (sayakpaul) [reply examples/image_classification-tf.ipynb:1]: I would love to contribute to it if possible. Are there open parts there?

Comment (Rocketknight1) [reply examples/image_classification-tf.ipynb:1]: Update (cc @amyeroberts): `pixel_values` should not be in the dataset because you didn't request it in `columns` or `label_cols`. It's only there because of an oversight in the recent `to_tf_dataset` code - it survives because it's added by the dataset transformation, and isn't a column in the original pre-transformation dataset. You definitely don't want to rely on that behaviour, though - `columns` and `label_cols` should contain the names of all the columns you want to see in the output!

Comment (amyeroberts) [reply examples/image_classification-tf.ipynb]: <p>Done!</p>

Comment (amyeroberts) [reply examples/image_classification-tf.ipynb]: <p>I've edited the metric callback to print out <code>val_accuracy</code>  at the end of epoch. This won't show train accuracy however. I believe this would require another callback to be added for computing the metric on train (is this right @Rocketknight1 ?) </p>

Comment (sayakpaul) [reply examples/image_classification-tf.ipynb]: Thanks!

Comment (amyeroberts) [reply examples/image_classification-tf.ipynb]: <p>Done! </p>

[the remark this answers is not in the document]
Comment (amyeroberts) [reply examples/image_classification-tf.ipynb]: <p>Left over from a copy/paste. Fixed now 👍</p>

**Comment by amyeroberts** [comment]
> @NielsRogge what's pending in this PR before it gets merged?

@sayakpaul One thing that is pending is a possible rework of the `to_tf_dataset` method. As you pointed out, it's confusing requesting columns which aren't actually in the resulting dataset (because of the applied transforms) cc @Rocketknight1. We can merge right away or wait for this. Normally, I'm pro iterating, but in this case, I'm not sure we want to show a demo with API calls that aren't intended.

**Comment by sayakpaul** [comment]
@amyeroberts since ResNets and RegNets will be compatible with `TFAutoModelForImageClassification` I think it'd be worthwhile to mention these two models (once ResNets are merged) too. Plus they are non-attention models like ConvNeXt (which you already mentioned in the notebook) so this might be helpful for the interested readers in choosing and experimenting with different kinds of models and architectures. 

WDYT?

**Comment by amyeroberts** [comment]
@sayakpaul Agreed! I'll add them :)

**Comment by sayakpaul** [comment]
@amyeroberts is there anything else that needs working for the PR to get merged?

**Comment by amyeroberts** [comment]
@sayakpaul So the only thing that's pending is the update that happened in `datasets` being part of a release so it's on main. I wonder if we could publish and require we're on dev in the notebook? WDYK @NielsRogge @Rocketknight1 ?

**Comment by Rocketknight1** [comment]
I'm fine with that! Alternatively, I think there'll be a release of `datasets` and `transformers` in the middle of next week, so you could just wait and merge it then. Either solution works for me.

**Comment by amyeroberts** [comment]
Added the changes you suggested @NielsRogge 

@Rocketknight1 @sayakpaul I'm going to merge as the changes are now on main for datasets (most recent release 5 days ago).

**Comment by amyeroberts** [comment]
@Rocketknight1 Not at all - thanks for implementing the patch!
